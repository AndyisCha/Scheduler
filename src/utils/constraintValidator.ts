import type { SlotConfig, ScheduleResult, TTScheduleResult, MWFScheduleResult } from '../types/scheduler';

export interface ConstraintViolation {
  type: 'unavailable' | 'homeroom_disabled' | 'max_homerooms' | 'missing_assignment';
  teacher: string;
  day: string;
  period: number;
  message: string;
  severity: 'warning' | 'error';
}

export interface ValidationResult {
  isValid: boolean;
  violations: ConstraintViolation[];
  warnings: string[];
}

// Overloaded function signatures
export function validateScheduleConstraints(
  slotConfig: SlotConfig,
  scheduleResult: ScheduleResult
): ValidationResult;
export function validateScheduleConstraints(
  slotConfig: SlotConfig,
  scheduleResult: TTScheduleResult
): ValidationResult;
export function validateScheduleConstraints(
  slotConfig: SlotConfig,
  scheduleResult: MWFScheduleResult
): ValidationResult;
export function validateScheduleConstraints(
  slotConfig: SlotConfig,
  scheduleResult: ScheduleResult | TTScheduleResult | MWFScheduleResult
): ValidationResult {
  const violations: ConstraintViolation[] = [];
  const warnings: string[] = [];

  const { slot } = slotConfig;
  const constraints = (slot as any).constraints || [];

  // Check each day and period in the schedule
  Object.entries(scheduleResult).forEach(([day, daySchedule]) => {
    Object.entries(daySchedule).forEach(([periodStr, assignmentData]) => {
      const period = parseInt(periodStr);
      
      // Handle both single assignment and array of assignments
      const assignments = Array.isArray(assignmentData) ? assignmentData : (assignmentData ? [assignmentData] : []);
      
      assignments.forEach(assignment => {
        if (assignment) {
          const teacherName = assignment.teacher;
          const constraint = constraints.find((c: any) => c.teacherName === teacherName);
          
          if (constraint) {
            // Check if teacher is scheduled during unavailable time
            const slotKey = `${day}|${period}`;
            if (constraint.unavailable.includes(slotKey)) {
              violations.push({
                type: 'unavailable',
                teacher: teacherName,
                day,
                period,
                message: `${teacherName} 교사가 불가능한 시간(${day} ${period}교시)에 배정되었습니다`,
                severity: 'error'
              });
            }

            // Check homeroom constraints
            if (assignment.role === 'H') {
              if (constraint.homeroomDisabled) {
                violations.push({
                  type: 'homeroom_disabled',
                  teacher: teacherName,
                  day,
                  period,
                  message: `${teacherName} 교사는 담임 불가인데 담임으로 배정되었습니다`,
                  severity: 'error'
                });
              }

              // Check max homerooms constraint
              const homeroomCount = countHomeroomAssignments(teacherName, scheduleResult);
              if (homeroomCount > constraint.maxHomerooms) {
                violations.push({
                  type: 'max_homerooms',
                  teacher: teacherName,
                  day,
                  period,
                  message: `${teacherName} 교사의 담임 배정이 최대 한도(${constraint.maxHomerooms})를 초과했습니다 (현재: ${homeroomCount})`,
                  severity: 'error'
                });
              }
            }
          }
        }
      });
    });
  });

  // Check for missing assignments
  const expectedClasses = calculateExpectedClasses(slotConfig);
  const actualAssignments = countActualAssignments(scheduleResult);
  
  if (actualAssignments < expectedClasses) {
    warnings.push(`예상 배정 수(${expectedClasses})보다 실제 배정 수(${actualAssignments})가 적습니다`);
  }

  return {
    isValid: violations.length === 0,
    violations,
    warnings
  };
}

function countHomeroomAssignments(teacherName: string, scheduleResult: ScheduleResult | TTScheduleResult | MWFScheduleResult): number {
  let count = 0;
  Object.values(scheduleResult).forEach(daySchedule => {
    Object.values(daySchedule).forEach(assignmentData => {
      if (Array.isArray(assignmentData)) {
        count += assignmentData.filter(assignment => assignment.teacher === teacherName && assignment.role === 'H').length;
      } else if (assignmentData && (assignmentData as any).teacher === teacherName && (assignmentData as any).role === 'H') {
        count++;
      }
    });
  });
  return count;
}

function calculateExpectedClasses(slotConfig: SlotConfig): number {
  const { slot } = slotConfig;
  const { globalOptions } = slot;
  
  let totalClasses = 0;
  Object.values(globalOptions.roundClassCounts).forEach(count => {
    totalClasses += count;
  });
  
  return totalClasses;
}

function countActualAssignments(scheduleResult: ScheduleResult | TTScheduleResult | MWFScheduleResult): number {
  let count = 0;
  Object.values(scheduleResult).forEach(daySchedule => {
    Object.values(daySchedule).forEach(assignmentData => {
      if (Array.isArray(assignmentData)) {
        count += assignmentData.filter(assignment => assignment).length;
      } else if (assignmentData) {
        count++;
      }
    });
  });
  return count;
}

export function generateConstraintViolationMessage(violations: ConstraintViolation[]): string {
  if (violations.length === 0) {
    return '모든 제약조건이 충족되었습니다';
  }

  const errorCount = violations.filter(v => v.severity === 'error').length;
  const warningCount = violations.filter(v => v.severity === 'warning').length;

  let message = `제약조건 위반이 발견되었습니다:\n`;
  message += `• 오류: ${errorCount}개\n`;
  message += `• 경고: ${warningCount}개\n\n`;
  
  violations.slice(0, 3).forEach(violation => {
    message += `• ${violation.message}\n`;
  });

  if (violations.length > 3) {
    message += `• ... 및 ${violations.length - 3}개 더`;
  }

  return message;
}
