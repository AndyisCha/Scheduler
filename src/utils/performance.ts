// 성능 최적화 유틸리티 함수들

/**
 * 함수를 디바운스하여 성능을 최적화합니다
 */
export function debounce<T extends (...args: any[]) => any>(
  func: T,
  wait: number,
  immediate = false
): (...args: Parameters<T>) => void {
  let timeout: NodeJS.Timeout | null = null;

  return function executedFunction(...args: Parameters<T>) {
    const later = () => {
      timeout = null;
      if (!immediate) func(...args);
    };

    const callNow = immediate && !timeout;

    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(later, wait);

    if (callNow) func(...args);
  };
}

/**
 * 함수를 스로틀하여 성능을 최적화합니다
 */
export function throttle<T extends (...args: any[]) => any>(
  func: T,
  limit: number
): (...args: Parameters<T>) => void {
  let inThrottle: boolean;

  return function executedFunction(...args: Parameters<T>) {
    if (!inThrottle) {
      func.apply(this, args);
      inThrottle = true;
      setTimeout(() => (inThrottle = false), limit);
    }
  };
}

/**
 * 객체의 깊은 비교를 수행합니다 (React.memo에서 사용)
 */
export function deepEqual(a: any, b: any): boolean {
  if (a === b) return true;

  if (a == null || b == null) return false;

  if (typeof a !== typeof b) return false;

  if (typeof a !== 'object') return false;

  const keysA = Object.keys(a);
  const keysB = Object.keys(b);

  if (keysA.length !== keysB.length) return false;

  for (const key of keysA) {
    if (!keysB.includes(key)) return false;
    if (!deepEqual(a[key], b[key])) return false;
  }

  return true;
}

/**
 * 배열의 중복을 제거합니다
 */
export function unique<T>(array: T[], key?: keyof T): T[] {
  if (!key) {
    return [...new Set(array)];
  }

  const seen = new Set();
  return array.filter(item => {
    const value = item[key];
    if (seen.has(value)) {
      return false;
    }
    seen.add(value);
    return true;
  });
}

/**
 * 배열을 청크로 나눕니다 (가상화에서 사용)
 */
export function chunk<T>(array: T[], size: number): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < array.length; i += size) {
    chunks.push(array.slice(i, i + size));
  }
  return chunks;
}

/**
 * 메모이제이션을 위한 간단한 캐시
 */
export function createMemoCache<T extends (...args: any[]) => any>(
  fn: T,
  maxSize = 100
): T {
  const cache = new Map<string, ReturnType<T>>();

  return ((...args: Parameters<T>): ReturnType<T> => {
    const key = JSON.stringify(args);
    
    if (cache.has(key)) {
      return cache.get(key)!;
    }

    const result = fn(...args);
    
    // 캐시 크기 제한
    if (cache.size >= maxSize) {
      const firstKey = cache.keys().next().value;
      cache.delete(firstKey);
    }
    
    cache.set(key, result);
    return result;
  }) as T;
}

/**
 * 비동기 작업의 결과를 캐시합니다
 */
export function createAsyncCache<T extends (...args: any[]) => Promise<any>>(
  fn: T,
  ttl = 5 * 60 * 1000 // 5분 기본값
): T {
  const cache = new Map<string, { result: any; timestamp: number }>();

  return (async (...args: Parameters<T>): Promise<ReturnType<T>> => {
    const key = JSON.stringify(args);
    const cached = cache.get(key);

    // 캐시된 결과가 있고 TTL이 만료되지 않았다면 반환
    if (cached && Date.now() - cached.timestamp < ttl) {
      return cached.result;
    }

    const result = await fn(...args);
    cache.set(key, { result, timestamp: Date.now() });
    
    return result;
  }) as T;
}

/**
 * 웹 워커를 사용하여 무거운 계산을 백그라운드에서 수행합니다
 */
export function createWebWorker<T, R>(
  workerScript: string
): (data: T) => Promise<R> {
  const worker = new Worker(
    URL.createObjectURL(new Blob([workerScript], { type: 'application/javascript' }))
  );

  return (data: T): Promise<R> => {
    return new Promise((resolve, reject) => {
      const handleMessage = (event: MessageEvent) => {
        worker.removeEventListener('message', handleMessage);
        worker.removeEventListener('error', handleError);
        resolve(event.data);
      };

      const handleError = (error: ErrorEvent) => {
        worker.removeEventListener('message', handleMessage);
        worker.removeEventListener('error', handleError);
        reject(error);
      };

      worker.addEventListener('message', handleMessage);
      worker.addEventListener('error', handleError);
      worker.postMessage(data);
    });
  };
}

/**
 * 가상화를 위한 인덱스 계산
 */
export function getVisibleRange(
  scrollTop: number,
  containerHeight: number,
  itemHeight: number,
  totalItems: number,
  overscan = 5
) {
  const startIndex = Math.max(0, Math.floor(scrollTop / itemHeight) - overscan);
  const endIndex = Math.min(
    totalItems - 1,
    Math.ceil((scrollTop + containerHeight) / itemHeight) + overscan
  );

  return { startIndex, endIndex };
}

/**
 * 성능 측정을 위한 마킹
 */
export function markPerformance(name: string, start = true) {
  if (typeof performance !== 'undefined' && performance.mark) {
    performance.mark(`${name}-${start ? 'start' : 'end'}`);
    
    if (!start) {
      try {
        performance.measure(name, `${name}-start`, `${name}-end`);
      } catch (error) {
        // 이미 측정된 경우 무시
      }
    }
  }
}

/**
 * 메모리 사용량 측정 (Chrome DevTools)
 */
export function getMemoryUsage() {
  if (typeof performance !== 'undefined' && (performance as any).memory) {
    const memory = (performance as any).memory;
    return {
      used: Math.round(memory.usedJSHeapSize / 1024 / 1024),
      total: Math.round(memory.totalJSHeapSize / 1024 / 1024),
      limit: Math.round(memory.jsHeapSizeLimit / 1024 / 1024)
    };
  }
  return null;
}
