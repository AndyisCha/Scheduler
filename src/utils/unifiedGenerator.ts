// src/utils/unifiedGenerator.ts
// Orchestrates MWF (existing engine) + TT (new engine) into a single weekly result.

import type { ScheduleResult as MWFResult, SchedulerSlot as MWFSlot } from "../types/scheduler"; // 기존 MWF 엔진 타입
import { createWeeklySchedule as createMWFWeekly } from "./scheduler";                      // 기존 MWF 엔진 함수
import type {
  TTSchedulerSlot, TTScheduleResult,
  FairnessLedger
} from "./ttScheduler";
import { createTTWeeklySchedule } from "./ttScheduler";

export type UnifiedOptions = {
  includeExams?: boolean; // (엔진 내부는 기본 포함 설계, 필요시 토글용)
  fairnessMode?: "balanced" | "F-priority" | "H-priority";
};

export type UnifiedWeekResult = {
  classSummary: Record<string, Record<"월"|"화"|"수"|"목"|"금", any[]>>;
  teacherSummary: Record<string, Record<"월"|"화"|"수"|"목"|"금", any[]>>;
  dayGrid: Record<"월"|"화"|"수"|"목"|"금", Record<number, any[]>>;
  warnings: string[];
  reports: {
    feasibilityTT: TTScheduleResult["feasibility"];
    fairness: {
      perTeacher: Record<string, { H: number; K: number; F: number; total: number }>;
      deviation: { H: number; K: number; F: number; total: number };
    };
  };
  mwfResult?: MWFResult;
  ttResult?: TTScheduleResult;
};

// 공정성 원장 초기화
function initLedger(): FairnessLedger {
  return {
    totalByTeacher: {},
    byRole: { H:{}, K:{}, F:{} },
  };
}

// MWF 결과를 ledger에 반영 (역할 카운트)
function absorbMWFIntoLedger(ledger: FairnessLedger, mwf: MWFResult) {
  for (const [teacher, days] of Object.entries(mwf.teacherSummary)) {
    for (const d of ["월","수","금"] as const) {
      const dayAssignments = (days as any)[d] || [];
      for (const sess of dayAssignments) {
        const role = sess.role as "H"|"K"|"F" | "EXAM";
        if (role === "EXAM") continue;
        ledger.totalByTeacher[teacher] = (ledger.totalByTeacher[teacher] || 0) + 1;
        ledger.byRole[role][teacher] = (ledger.byRole[role][teacher] || 0) + 1;
      }
    }
  }
}

// TT 결과를 ledger에 반영하는 것은 ttScheduler가 내부에서 bump 하므로 필요 없음.
// 하지만 합치기 전/후 카운트 집계는 리포트에서 계산.

export function generateUnifiedWeek(
  mwfSlot: MWFSlot,
  ttSlot: TTSchedulerSlot,
  _opts: UnifiedOptions = {}
): UnifiedWeekResult {
  const ledger = initLedger();

  // 1) MWF 생성 (월/수/금)
  const mwf = createMWFWeekly({ ...mwfSlot }); // 기존 엔진: 내부에서 자체 카운트만 하므로, ledger에 반영 필요 → absorb
  absorbMWFIntoLedger(ledger, mwf);

  // 2) TT 생성 (화/목) — 공유 ledger로 공정성 이어받음
  const tt = createTTWeeklySchedule(ttSlot, ledger, ["화","목"]);

  // 3) 병합
  const unified: UnifiedWeekResult = {
    classSummary: {},
    teacherSummary: {},
    dayGrid: {
      "월": mwf.dayGrid?.["월"] || {},
      "화": tt.dayGrid?.["화"] || {},
      "수": mwf.dayGrid?.["수"] || {},
      "목": tt.dayGrid?.["목"] || {},
      "금": mwf.dayGrid?.["금"] || {},
    } as any,
    warnings: [],
    reports: {
      feasibilityTT: tt.feasibility,
      fairness: { perTeacher: {}, deviation: { H:0, K:0, F:0, total:0 } }
    }
  };

  // classSummary 병합
  // MWF: mwf.classSummary -> 월/수/금
  for (const [cid, days] of Object.entries(mwf.classSummary)) {
    unified.classSummary[cid] ||= { "월":[], "화":[], "수":[], "목":[], "금":[] } as any;
    unified.classSummary[cid]["월"] = days["월"] || [];
    unified.classSummary[cid]["수"] = days["수"] || [];
    unified.classSummary[cid]["금"] = days["금"] || [];
  }
  // TT: tt.classSummary -> 화/목
  for (const [cid, days] of Object.entries(tt.classSummary)) {
    unified.classSummary[cid] ||= { "월":[], "화":[], "수":[], "목":[], "금":[] } as any;
    unified.classSummary[cid]["화"] = days["화"] || [];
    unified.classSummary[cid]["목"] = days["목"] || [];
  }

  // teacherSummary 병합
  for (const [t, days] of Object.entries(mwf.teacherSummary)) {
    unified.teacherSummary[t] ||= { "월":[], "화":[], "수":[], "목":[], "금":[] } as any;
    unified.teacherSummary[t]["월"] = days["월"] || [];
    unified.teacherSummary[t]["수"] = days["수"] || [];
    unified.teacherSummary[t]["금"] = days["금"] || [];
  }
  for (const [t, days] of Object.entries(tt.teacherSummary)) {
    unified.teacherSummary[t] ||= { "월":[], "화":[], "수":[], "목":[], "금":[] } as any;
    unified.teacherSummary[t]["화"] = days["화"] || [];
    unified.teacherSummary[t]["목"] = days["목"] || [];
  }

  // warnings
  unified.warnings.push(...(mwf.warnings || []));
  unified.warnings.push(...(tt.warnings || []));

  // 4) 공정성 리포트 (주간 H/K/F/total per teacher)
  const per: Record<string, { H:number; K:number; F:number; total:number }> = {};
  const add = (teacher: string, role: "H"|"K"|"F") => {
    per[teacher] ||= { H:0, K:0, F:0, total:0 };
    per[teacher][role]++; per[teacher].total++;
  };
  // from teacherSummary merged
  for (const [t, days] of Object.entries(unified.teacherSummary)) {
    for (const d of ["월","화","수","목","금"] as const) {
      for (const sess of (days[d] || [])) {
        const r = sess.role as "H"|"K"|"F"|"EXAM";
        if (r === "EXAM") continue;
        add(t, r);
      }
    }
  }
  // deviation
  const dev = (k: "H"|"K"|"F"|"total") => {
    const vals = Object.values(per).map(v => v[k]).sort((a,b)=>a-b);
    if (vals.length === 0) return 0;
    return vals[vals.length-1] - vals[0];
  };
  unified.reports.fairness.perTeacher = per;
  unified.reports.fairness.deviation = {
    H: dev("H"), K: dev("K"), F: dev("F"), total: dev("total")
  };

  // Include original results for history service
  unified.mwfResult = mwf;
  unified.ttResult = tt;

  return unified;
}
