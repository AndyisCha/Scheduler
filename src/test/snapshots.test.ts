// Snapshot tests for key outputs: warnings, feasibility, fairness reports
import { describe, it, expect } from 'vitest'
import { generateUnifiedWeek } from '../utils/unifiedGenerator'
import { createTTWeeklySchedule } from '../utils/ttScheduler'
import { ScenarioFactories } from './factories'

describe('Snapshot Tests - Key Outputs', () => {
  
  describe('Scenario A: Optimal case - warnings snapshot', () => {
    it('should produce minimal warnings for feasible scenario', () => {
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'balanced'
      })

      // Snapshot warnings (should be minimal for optimal case)
      expect(result.warnings).toMatchSnapshot('scenario-a-warnings')
    })
  })

  describe('Scenario B: Infeasible case - warnings snapshot', () => {
    it('should produce critical warnings for infeasible scenario', () => {
      const { mwf, tt } = ScenarioFactories.scenarioB()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'F-priority'
      })

      // Snapshot warnings (should include capacity warnings)
      expect(result.warnings).toMatchSnapshot('scenario-b-warnings')
    })
  })

  describe('Scenario D: Korean shortage - warnings snapshot', () => {
    it('should produce Korean teacher shortage warnings', () => {
      const { mwf, tt } = ScenarioFactories.scenarioD()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'H-priority'
      })

      // Snapshot warnings (should include Korean unavailability warnings)
      expect(result.warnings).toMatchSnapshot('scenario-d-warnings')
    })
  })

  describe('TT Feasibility Reports - snapshots', () => {
    it('should produce consistent feasibility reports for scenario A', () => {
      const { tt } = ScenarioFactories.scenarioA()
      
      const ttResult = createTTWeeklySchedule(
        tt, 
        { totalByTeacher: {}, byRole: { H: {}, K: {}, F: {} } }, 
        ['화', '목']
      )

      // Snapshot TT feasibility report
      expect(ttResult.feasibility).toMatchSnapshot('scenario-a-tt-feasibility')
    })

    it('should produce consistent feasibility reports for scenario B', () => {
      const { tt } = ScenarioFactories.scenarioB()
      
      const ttResult = createTTWeeklySchedule(
        tt, 
        { totalByTeacher: {}, byRole: { H: {}, K: {}, F: {} } }, 
        ['화', '목']
      )

      // Snapshot TT feasibility report (should show infeasibility)
      expect(ttResult.feasibility).toMatchSnapshot('scenario-b-tt-feasibility')
    })

    it('should produce consistent feasibility reports for scenario D', () => {
      const { tt } = ScenarioFactories.scenarioD()
      
      const ttResult = createTTWeeklySchedule(
        tt, 
        { totalByTeacher: {}, byRole: { H: {}, K: {}, F: {} } }, 
        ['화', '목']
      )

      // Snapshot TT feasibility report (should show Korean capacity issues)
      expect(ttResult.feasibility).toMatchSnapshot('scenario-d-tt-feasibility')
    })
  })

  describe('Fairness Reports - snapshots', () => {
    it('should produce consistent fairness reports for scenario A', () => {
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'balanced'
      })

      // Snapshot fairness report
      expect(result.reports.fairness).toMatchSnapshot('scenario-a-fairness')
    })

    it('should produce consistent fairness reports for scenario E', () => {
      const { mwf, tt } = ScenarioFactories.scenarioE()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'balanced'
      })

      // Snapshot fairness report (should show good balance)
      expect(result.reports.fairness).toMatchSnapshot('scenario-e-fairness')
    })

    it('should produce consistent fairness reports for F-priority mode', () => {
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'F-priority'
      })

      // Snapshot fairness report (should prioritize foreign assignments)
      expect(result.reports.fairness).toMatchSnapshot('scenario-a-f-priority-fairness')
    })

    it('should produce consistent fairness reports for H-priority mode', () => {
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'H-priority'
      })

      // Snapshot fairness report (should prioritize homeroom assignments)
      expect(result.reports.fairness).toMatchSnapshot('scenario-a-h-priority-fairness')
    })
  })

  describe('Deterministic Output - same inputs produce same outputs', () => {
    it('should produce identical results for multiple runs of scenario A', () => {
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      // Run multiple times to ensure deterministic output
      const results = []
      for (let i = 0; i < 3; i++) {
        const result = generateUnifiedWeek(mwf, tt, {
          includeExams: true,
          fairnessMode: 'balanced'
        })
        results.push(result)
      }

      // All results should be identical
      const firstResult = results[0]
      for (let i = 1; i < results.length; i++) {
        expect(results[i].warnings).toEqual(firstResult.warnings)
        expect(results[i].reports.feasibilityTT).toEqual(firstResult.reports.feasibilityTT)
        expect(results[i].reports.fairness).toEqual(firstResult.reports.fairness)
      }
    })
  })

  describe('Small Fixture Validation', () => {
    it('should use minimal test data for consistent snapshots', () => {
      // Verify our test fixtures are minimal and deterministic
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      // MWF slot should have minimal teacher pool
      expect(mwf.teachers.homeroomKoreanPool).toHaveLength(4)
      expect(mwf.teachers.foreignPool).toHaveLength(1)
      expect(Object.keys(mwf.teachers.constraints)).toHaveLength(5) // 4 Korean + 1 foreign
      
      // TT slot should have minimal configuration
      expect(tt.teachers.homeroomKoreanPool).toHaveLength(4)
      expect(tt.teachers.foreignPool).toHaveLength(1)
      expect(tt.globalOptions.roundClassCounts).toEqual({ 1: 3, 2: 3 })
      
      // Verify no extra complexity
      expect(mwf.fixedHomerooms).toEqual({})
      expect(tt.fixedHomerooms).toEqual({})
    })

    it('should produce compact snapshot data', () => {
      const { mwf, tt } = ScenarioFactories.scenarioA()
      
      const result = generateUnifiedWeek(mwf, tt, {
        includeExams: true,
        fairnessMode: 'balanced'
      })

      // Warnings should be minimal for optimal case
      expect(result.warnings.length).toBeLessThanOrEqual(5)
      
      // Fairness report should be compact
      const teacherCount = Object.keys(result.reports.fairness.perTeacher).length
      expect(teacherCount).toBeLessThanOrEqual(6) // Allow for 4 Korean + 1 foreign + potential extras
      
      // Feasibility should be simple
      expect(typeof result.reports.feasibilityTT.r1ForeignOk).toBe('boolean')
      expect(typeof result.reports.feasibilityTT.r1ForeignDemand).toBe('number')
      expect(typeof result.reports.feasibilityTT.r1ForeignCapacity).toBe('number')
    })
  })
})
