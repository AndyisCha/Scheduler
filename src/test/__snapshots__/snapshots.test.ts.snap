// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot Tests - Key Outputs > Fairness Reports - snapshots > should produce consistent fairness reports for F-priority mode > scenario-a-f-priority-fairness 1`] = `
{
  "deviation": {
    "F": 18,
    "H": 10,
    "K": 7,
    "total": 14,
  },
  "perTeacher": {
    "(미배정)": {
      "F": 0,
      "H": 4,
      "K": 0,
      "total": 4,
    },
    "John": {
      "F": 18,
      "H": 0,
      "K": 0,
      "total": 18,
    },
    "김선생": {
      "F": 0,
      "H": 10,
      "K": 7,
      "total": 17,
    },
    "박선생": {
      "F": 0,
      "H": 9,
      "K": 7,
      "total": 16,
    },
    "이선생": {
      "F": 0,
      "H": 6,
      "K": 7,
      "total": 13,
    },
    "최선생": {
      "F": 0,
      "H": 9,
      "K": 7,
      "total": 16,
    },
  },
}
`;

exports[`Snapshot Tests - Key Outputs > Fairness Reports - snapshots > should produce consistent fairness reports for H-priority mode > scenario-a-h-priority-fairness 1`] = `
{
  "deviation": {
    "F": 18,
    "H": 10,
    "K": 7,
    "total": 14,
  },
  "perTeacher": {
    "(미배정)": {
      "F": 0,
      "H": 4,
      "K": 0,
      "total": 4,
    },
    "John": {
      "F": 18,
      "H": 0,
      "K": 0,
      "total": 18,
    },
    "김선생": {
      "F": 0,
      "H": 10,
      "K": 7,
      "total": 17,
    },
    "박선생": {
      "F": 0,
      "H": 9,
      "K": 7,
      "total": 16,
    },
    "이선생": {
      "F": 0,
      "H": 6,
      "K": 7,
      "total": 13,
    },
    "최선생": {
      "F": 0,
      "H": 9,
      "K": 7,
      "total": 16,
    },
  },
}
`;

exports[`Snapshot Tests - Key Outputs > Fairness Reports - snapshots > should produce consistent fairness reports for scenario A > scenario-a-fairness 1`] = `
{
  "deviation": {
    "F": 18,
    "H": 10,
    "K": 7,
    "total": 14,
  },
  "perTeacher": {
    "(미배정)": {
      "F": 0,
      "H": 4,
      "K": 0,
      "total": 4,
    },
    "John": {
      "F": 18,
      "H": 0,
      "K": 0,
      "total": 18,
    },
    "김선생": {
      "F": 0,
      "H": 10,
      "K": 7,
      "total": 17,
    },
    "박선생": {
      "F": 0,
      "H": 9,
      "K": 7,
      "total": 16,
    },
    "이선생": {
      "F": 0,
      "H": 6,
      "K": 7,
      "total": 13,
    },
    "최선생": {
      "F": 0,
      "H": 9,
      "K": 7,
      "total": 16,
    },
  },
}
`;

exports[`Snapshot Tests - Key Outputs > Fairness Reports - snapshots > should produce consistent fairness reports for scenario E > scenario-e-fairness 1`] = `
{
  "deviation": {
    "F": 8,
    "H": 10,
    "K": 8,
    "total": 13,
  },
  "perTeacher": {
    "H-R3C3": {
      "F": 0,
      "H": 2,
      "K": 2,
      "total": 4,
    },
    "H-R4C1": {
      "F": 0,
      "H": 4,
      "K": 2,
      "total": 6,
    },
    "H-R4C2": {
      "F": 0,
      "H": 4,
      "K": 2,
      "total": 6,
    },
    "H-R4C3": {
      "F": 0,
      "H": 4,
      "K": 1,
      "total": 5,
    },
    "John": {
      "F": 8,
      "H": 0,
      "K": 0,
      "total": 8,
    },
    "Mike": {
      "F": 8,
      "H": 0,
      "K": 0,
      "total": 8,
    },
    "Sarah": {
      "F": 8,
      "H": 0,
      "K": 0,
      "total": 8,
    },
    "김선생": {
      "F": 0,
      "H": 10,
      "K": 7,
      "total": 17,
    },
    "박선생": {
      "F": 0,
      "H": 10,
      "K": 7,
      "total": 17,
    },
    "이선생": {
      "F": 0,
      "H": 6,
      "K": 8,
      "total": 14,
    },
    "최선생": {
      "F": 0,
      "H": 8,
      "K": 7,
      "total": 15,
    },
  },
}
`;

exports[`Snapshot Tests - Key Outputs > Scenario A: Optimal case - warnings snapshot > should produce minimal warnings for feasible scenario > scenario-a-warnings 1`] = `
[
  "[금 2] R1C2 H 배정 실패",
  "[금 4] R2C2 H 배정 실패",
  "[화 2] R1C3 R1 H 배정 실패",
  "[목 2] R1C3 R1 H 배정 실패",
]
`;

exports[`Snapshot Tests - Key Outputs > Scenario B: Infeasible case - warnings snapshot > should produce critical warnings for infeasible scenario > scenario-b-warnings 1`] = `
[
  "[금 2] R1C2 H 배정 실패",
  "[금 4] R2C2 H 배정 실패",
  "[화 2] R1C3 R1 H 배정 실패",
  "[화 3] R1C4 R1 F 배정 실패",
  "[목 2] R1C3 R1 H 배정 실패",
  "[목 3] R1C4 R1 F 배정 실패",
]
`;

exports[`Snapshot Tests - Key Outputs > Scenario D: Korean shortage - warnings snapshot > should produce Korean teacher shortage warnings > scenario-d-warnings 1`] = `
[
  "[화 4] R2C1 R2 H 배정 실패",
  "[화 5] R2C1 R2 H 배정 실패",
  "[화 6] R2C2 R2 K 배정 실패",
  "[화 4] R2C3 R2 H 배정 실패",
  "[화 5] R2C3 R2 H 배정 실패",
  "[화 6] R2C3 R2 K 배정 실패",
  "[목 4] R2C1 R2 H 배정 실패",
  "[목 5] R2C1 R2 H 배정 실패",
  "[목 6] R2C2 R2 K 배정 실패",
  "[목 4] R2C3 R2 H 배정 실패",
  "[목 5] R2C3 R2 H 배정 실패",
  "[목 6] R2C3 R2 K 배정 실패",
]
`;

exports[`Snapshot Tests - Key Outputs > TT Feasibility Reports - snapshots > should produce consistent feasibility reports for scenario A > scenario-a-tt-feasibility 1`] = `
{
  "r1ForeignCapacity": 6,
  "r1ForeignDemand": 6,
  "r1ForeignOk": true,
  "r2HNeeded": 12,
  "r2KNeeded": 6,
}
`;

exports[`Snapshot Tests - Key Outputs > TT Feasibility Reports - snapshots > should produce consistent feasibility reports for scenario B > scenario-b-tt-feasibility 1`] = `
{
  "r1ForeignCapacity": 6,
  "r1ForeignDemand": 8,
  "r1ForeignOk": false,
  "r2HNeeded": 12,
  "r2KNeeded": 6,
}
`;

exports[`Snapshot Tests - Key Outputs > TT Feasibility Reports - snapshots > should produce consistent feasibility reports for scenario D > scenario-d-tt-feasibility 1`] = `
{
  "r1ForeignCapacity": 18,
  "r1ForeignDemand": 6,
  "r1ForeignOk": true,
  "r2HNeeded": 12,
  "r2KNeeded": 6,
}
`;
